### ISSUE TITLE
Render to regex syntax

### ISSUE THREAD

**Issue by chrisdone** [feuerbach/regex-applicative#36]
I'm making a new form library like formlets, and considering the ability to embed a regex as a validator/parser which would run on the server side with regex-applicative upon submission, but _also_, as a UX optimization, compile the same regex to a `[a-z]{2}+`-style regex to be used as the `pattern=".."` [attribute for html5](https://html.spec.whatwg.org/multipage/input.html#the-pattern-attribute) (avoiding a roundtrip just to do a simple text format check). This would enable a developer to build forms with client and server-side validation composed using the same regex-applicative recipe.

So the question is: is there already a library or module that would convert an `RE s a` to `Text`? E.g.

```haskell
convert (many (sym 'a')) => "a*"
```

I assume I could write my own, but perhaps it already exists.

Thanks!

**Comment by feuerbach** [comment]
Unfortunately, this is not possible because `sym 'a'` is translated to `psym (== 'a')`, and you cannot extract `'a'` out of `(== 'a')`.

**Comment by chrisdone** [comment]
@feuerbach Ah, ok. So it was worth asking. Thanks.

I suppose then I would need to wrap up the API with an interface which doesn't expose `psym`. E.g.

```
data Regex a
```

with a `Regex a -> RE Char a` conversion function. Whipping up a simple GADT that is the same as the `RE` one but without this capability should be trivial, and I can re-use the rest of the library.

If you're interested, I could send it as a PR against this package as e.g. `Text.Regex.Applicative.Renderable` or w/e, otherwise package it as `regex-applicative-renderable` or something. I'm sure I'll find other uses than just this form package, and the work is probably trivial.

**Comment by feuerbach** [comment]
Yeah, or maybe have a class containing `sym` and perhaps some other functions, and make both types instances of that class.

The other constructors it should probably support are:

1. Recognizing a set (as in `Data.Set.Set`) of characters — can be simulated by summing individual `sym`s, but could be implemented more efficiently.
1. Recognizing a class (as in a POSIX regex class — `alpha`, `alnum` etc.) of characters — can in theory be implemented via individual syms, but in practice would be prohibitive.

This new module/representation might also support more efficient compilation, i.e. it's potential utility won't be restricted just to rendering. Let's call it `Text.Regex.Applicative.Char`?

**Comment by chrisdone** [comment]
OK, great then. When I get to this part of my lib I'll circle back to this issue. Thanks, Roman.

**Comment by treeowl** [comment]
A Set of characters will perform considerably worse, for almost all
operations, than an IntMap of their equivalents.

**Comment by chrisdone** [comment]
@treeowl do you have benchmarks for that?

**Comment by ceedubs** [comment]
I stumbled upon this discussion today. I had written a Scala regex library that a while ago I rewrote to more or less be a port of regex-applicative. The way I handled this in my library is that the the equivalent of the `Symbol` constructor (I call it [Elem](https://github.com/ceedubs/irrec/blob/master/regex/src/main/scala/Regex.scala#L67-L69)) has a separate `metadata` field of an abstract type `m` (which ends up needing to be another type parameter on `RE`).

For character (and any other discrete type) regexes, the `m` type ends up being [a type](https://github.com/ceedubs/irrec/blob/master/regex/src/main/scala/Match.scala) that looks something like:

```haskell
data Match a = Literal a | Wildcard | Allow (Diet a) | Forbid (Diet a)
```

where `Diet` is a discrete interval tree. This represents the notion of character classes in traditional regular expressions, supporting unions and intersections. You can then write [helper methods](https://github.com/ceedubs/irrec/blob/150955a3265c2f2efe96ddc03fbd426b1f333d66/regex/src/main/scala/Match.scala#L10-L16) that implement the `a -> maybe b` function based on the `Match`.

In addition to being able to get a string representation of the regex, a nice benefit of having this metadata around is that you can then convert a regular expression to a Quickcheck generator that generates values that match that regex. I have a demo page [here](https://ceedubs.github.io/irrec/docs/regex-explorer.html) that I think is kind of fun.

Now is probably a good time to say thanks for creating this library, @feuerbach! As soon as I saw it I knew that I wanted to rewrite my Scala library :)

**Comment by feuerbach** [comment]
Thanks @ceedubs, those are some nice ideas.
